### DISCUSSION TITLE
Postgres connection from serverless function

### DISCUSSION THREAD

**Discussion by holgerw** [vercel/next.js#13725] [Help]
My question is how to guarantee that the number of DB connections from server-side code (api routes, getServerSideProps) is kept below the given limit of a Postgres server? My question referes to a production cenário with serverless functions on the Vercel platform.

I studied the question and found 2 different parcial solutions: A) implementing a connection retry or B) limiting the connection pool of a single serverless function instance  and guarantee that this poll is reused among function invocations during the lifetime of the NodeJS process. 

A)
https://vercel.com/guides/deploying-next-and-mysql-with-vercel 
https://github.com/jeremydaly/serverless-mysql
https://github.com/MatteoGioioso/serverless-pg

B) 
https://gist.github.com/streamich/6175853840fb5209388405910c6cc04b

To my understanding solutions A) do not guarantee that a connection will be acquired with the retry strategy, potentially resulting in a failed request. And at least the retry backoff introduces request latency. And solution B) does limit the number of connections scoped only to a single nodeJS process (serverless function instance), resulting in connection errors when function instances are scaled unlimited.

In my understanding the only bullet-proof solution would be limiting the connection pool of a single serverless function instance (solution B above) together with a control/limit of the serverless function scaling by configuration.  Does this exist?
Note that limiting the serverless function scaling might make sense even from a performance point of view, as queueing requests on the webserver socket might be better then to have sufficient function instances for parallel processing and then queueing at the DB connection retry mechanism proposed by solutions A). 

So my fundamental question: Is there any bullet-proof approach to control DB connection behaviour when deploying Next.js to Vercel's serverless functions? Or is it better do desist?

If there is a solution, I think it would be really valuable to have it officially documented and endorsed. Right now Next.js promise of rapid development and Vercel's promise of rapid transition to deployment is hurt when needing to connect to relational DBs, because of the amount of research and experimentation necessary to solve this. And even then we push to production with doubt if the adopted solution is really bullet-proof and won't throw exceptions under load.

I really appreciate the Next.js maintainers efforts and also the communities dedication in order to make this cool technology. Please forgive any negative tone above. It is due to the frustration with something really basic like connecting  reliably to a DB. But I appreciate the effort of everyone involved.

**Comment by holgerw** [comment]
One solution I also see is to deploy Next.js as a traditional Node application with "next start". In this case we can exactly control the number of DB connections. It is great that Next.js supports this form of deployments to escape any current limitations of serverless environments.

Still interested if there is a bullet-proof solution to the problem when deploying to serverless. It would be great to deploy to serverless to benefit from Vercel's value proposition of deployment without hurdles.

**Comment by filipedeschamps** [comment]
> One solution I also see is to deploy Next.js as a traditional Node application

As far as I see the evolution of the framework, I wouldn't invest in that route.

But I'm curious, have you found a reliable way to "infinite" scale?

Best regards.

**Comment by holgerw** [comment]
Thanks for our reply.

One solution I also see is to deploy Next.js as a traditional Node
> application
>
> As far as I see the evolution of the framework, I wouldn't invest in that
> route.
>
Regarding the ability to run a built Next.js site as a Node app on a self
maintained server, do you remember any statement or sign that this may not
be supported further?

> But I'm curious, have you found a reliable way to "infinite" scale?
>
Recently AWS launched a DB proxy as a managed service. This may be efective
to prevent connection errors. Maybe connection latency is good enough also
if hosted in the same region as Vercel workers.

https://aws.amazon.com/pt/rds/proxy/

Best regards

**Reply by filipedeschamps** [reply]
About the custom server: https://spectrum.chat/next-js/general/is-the-custom-server-feature-of-next-js-going-to-be-deprecated~0cf21d56-4313-42e5-9be9-8114ae6c01a1

Specially this @Timer reply:

![image](https://user-images.githubusercontent.com/4248081/92315758-74678000-ef9f-11ea-90e4-fa400d6bc5ad.png)

I don't know if anything changed about this.

**Reply by jperl** [reply]
RDS proxy is not accessible outside of an AWS VPC

**Reply by kripod** [reply]
I've just asked if RDS Proxy could be accessed through the public internet on the AWS forums: https://forums.aws.amazon.com/thread.jspa?threadID=332344

**Reply by jgoux** [reply]
Hey @kripod, I saw that you didn't receive any answer on AWS forums, did you find a solution to make RDS Proxy public?

**Reply by kripod** [reply]
Hello,
Thank you for asking. Unfortunately, I haven’t found a solution but noticed that Vercel has a Serverless function execution limit of 1000, which doesn’t exceed the amount of connections the AWS DB can handle without a proxy.

**Reply by leerob** [reply]
One thing to clarify here, as I see it quite a bit. Custom Server !== using `next start` in a serverful environment. We still recommend using `next start`, whether that's self-hosted (serverful) or serverless (Vercel)!

**Comment by ivan-kleshnin** [comment]
@holgerw good post, I also invested a significant amount of time to this question with no clear answers.
To my understanding, serverless-specific library is the way to go but there's none looking at least partially ready 😞 
It freaks me out to use something like `pg-serveless` with 7 ⭐  in production.

The approach I use at the moment is a cloud-based pool https://www.digitalocean.com/docs/databases/postgresql/how-to/manage-connection-pools With such pool, when the number of DB connections from NextJS grows at scale, the number of **real** db connections stays limited.

**Reply by leerob** [reply]
We've since added some of this guidance to our docs: https://vercel.com/docs/solutions/databases
